File: .skills/openclaw-skills/skills/echoudhry/clawculator/htmlReport.js
'use strict';

const fs = require('fs');
const path = require('path');
const os = require('os');

function severityColor(severity) {
  return { critical: '#ef4444', high: '#f97316', medium: '#eab308', low: '#06b6d4', info: '#22c55e' }[severity] || '#6b7280';
}

function severityBg(severity) {
  return { critical: '#fef2f2', high: '#fff7ed', medium: '#fefce8', low: '#ecfeff', info: '#f0fdf4' }[severity] || '#f9fafb';
}

function severityIcon(severity) {
  return { critical: '🔴', high: '🟠', medium: '🟡', low: '🔵', info: '✅' }[severity] || '⚪';
}

function relativeAge(ageMs) {
  if (!ageMs) return 'unknown';
  const s = Math.floor(ageMs / 1000);
  if (s < 60)        return `${s}s ago`;
  const m = Math.floor(s / 60);
  if (m < 60)        return `${m}m ago`;
  const h = Math.floor(m / 60);
  if (h < 24)        return `${h}h ago`;
  const d = Math.floor(h / 24);
  if (d < 30)        return `${d}d ago`;
  return `${Math.floor(d / 30)}mo ago`;
}

const SOURCE_LABELS = {
  heartbeat: '💓 Heartbeat', hooks: '🪝 Hooks', whatsapp: '📱 WhatsApp',
  subagents: '🤖 Subagents', skills: '🔧 Skills', memory: '🧠 Memory',
  primary_model: '⚙️ Primary Model', sessions: '💬 Sessions',
  workspace: '📁 Workspace', config: '📄 Config',
};

async function generateHTMLReport(analysis, outPath) {
  const { summary, findings, sessions } = analysis;
  const bleed = summary.estimatedMonthlyBleed;

  // Burn summary calculation — only tracked sessions (not historical untracked)
  const trackedSessions = (sessions || []).filter(s => !s.isUntracked && s.dailyCost);
  const totalDaily = trackedSessions.reduce((sum, s) => sum + (s.dailyCost || 0), 0);
  const totalMonthly = totalDaily * 30;
  const burnSummary = totalDaily > 0 ? `
    <div class="section">
      <div class="section-title">Session Burn Rate</div>
      <div style="display:flex; gap:32px; flex-wrap:wrap;">
        <div>
          <div style="font-size:24px; font-weight:800; color:#f59e0b;">$${totalDaily.toFixed(4)}/day</div>
          <div style="font-size:13px; color:#94a3b8; margin-top:4px;">Combined daily burn rate across ${trackedSessions.length} active session${trackedSessions.length !== 1 ? 's' : ''}</div>
        </div>
        <div>
          <div style="font-size:24px; font-weight:800; color:#f97316;">$${totalMonthly.toFixed(2)}/month</div>
          <div style="font-size:13px; color:#94a3b8; margin-top:4px;">Projected monthly from active sessions</div>
        </div>
      </div>
    </div>` : '';

  const findingCards = findings.map(f => `
    <div class="finding" style="border-left: 4px solid ${severityColor(f.severity)}; background: ${severityBg(f.severity)}; padding: 16px; margin-bottom: 12px; border-radius: 0 8px 8px 0;">
      <div style="display:flex; align-items:center; gap:8px; margin-bottom:6px;">
        <span style="font-size:18px">${severityIcon(f.severity)}</span>
        <strong style="color:${severityColor(f.severity)}">${f.severity.toUpperCase()}</strong>
        <span style="color:#6b7280; font-size:14px">${SOURCE_LABELS[f.source] || f.source}</span>
        ${f.monthlyCost ? `<span style="margin-left:auto; color:${severityColor(f.severity)}; font-weight:bold">$${f.monthlyCost.toFixed(2)}/mo</span>` : ''}
      </div>
      <div style="font-weight:600; color:#111; margin-bottom:4px">${f.message}</div>
      ${f.detail ? `<div style="color:#555; font-size:14px; margin-bottom:6px; white-space:pre-line">${f.detail}</div>` : ''}
      ${f.fix ? `<div style="color:#16a34a; font-size:14px; margin-top:8px;">→ ${f.fix}</div>` : ''}
      ${f.command ? `<div style="background:#0f172a; color:#38bdf8; font-family:monospace; font-size:13px; padding:10px 14px; border-radius:6px; margin-top:6px; white-space:pre-wrap; border:1px solid #1e293b;"><span style="color:#6b7280; user-select:none;">$ </span>${f.command}</div>` : ''}
    </div>
  `).join('');

  const makeSessionRow = (s) => {
      const keyDisplay = s.key.length > 12 ? s.key.slice(0, 8) + '…' : s.key;
      const flag = s.isOrphaned ? ' ⚠️' : '';
      const txBadge = s.hasTranscript ? '<span style="color:#22c55e; font-size:10px; margin-left:4px" title="Cost from .jsonl transcript (API-reported)">●</span>' : '<span style="color:#6b7280; font-size:10px; margin-left:4px" title="Estimated cost (no transcript found)">○</span>';
      const age = s.ageMs ? relativeAge(s.ageMs) : 'unknown';
      const absDate = s.updatedAt ? new Date(s.updatedAt).toLocaleString() : '';
      const daily = s.dailyCost ? `$${s.dailyCost.toFixed(4)}/day` : '—';
      const cacheTitle = (s.cacheRead || s.cacheWrite) ? `Cache R: ${(s.cacheRead||0).toLocaleString()} · W: ${(s.cacheWrite||0).toLocaleString()}` : '';
      const rowBg = s.isOrphaned ? 'background:#fff7ed' : (s.isRecent ? 'background:#fffbeb' : '');
      return `
      <tr style="${rowBg}">
        <td style="padding:8px 12px; font-family:monospace; font-size:13px">${keyDisplay}${flag}${txBadge}</td>
        <td style="padding:8px 12px">${s.modelLabel || s.model}</td>
        <td style="padding:8px 12px; text-align:right" title="${cacheTitle}">${(s.inputTokens + s.outputTokens).toLocaleString()}${(s.cacheRead || s.cacheWrite) ? `<span style="color:#6b7280; font-size:11px;"> +${((s.cacheRead||0)+(s.cacheWrite||0)).toLocaleString()} cache</span>` : ''}</td>
        <td style="padding:8px 12px; text-align:right; color:${s.cost > 0.01 ? '#ef4444' : '#22c55e'}">$${s.cost.toFixed(4)}</td>
        <td style="padding:8px 12px; text-align:right; color:#f59e0b">${daily}</td>
        <td style="padding:8px 12px; color:#6b7280; font-size:13px" title="${absDate}">${age}</td>
      </tr>`;
  };

  const activeSessions = (sessions || []).filter(s => !s.isUntracked);
  const untrackedSessions = (sessions || []).filter(s => s.isUntracked);

  const activeRows = activeSessions
    .sort((a, b) => b.cost - a.cost)
    .slice(0, 20)
    .map(makeSessionRow).join('');

  const untrackedRows = untrackedSessions
    .sort((a, b) => b.cost - a.cost)
    .slice(0, 10)
    .map(makeSessionRow).join('');

  const untrackedTotal = untrackedSessions.reduce((sum, s) => sum + s.cost, 0);
  const untrackedHidden = Math.max(0, untrackedSessions.length - 10);

  const html = `<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <title>Clawculator Report — ${new Date(analysis.scannedAt).toLocaleString()}</title>
  <style>
    * { box-sizing: border-box; margin: 0; padding: 0; }
    body { font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', sans-serif; background: #0f172a; color: #e2e8f0; min-height: 100vh; }
    .header { background: linear-gradient(135deg, #0f172a 0%, #1e3a5f 100%); padding: 48px 32px; text-align: center; border-bottom: 1px solid #1e40af; }
    .logo { font-size: 42px; font-weight: 900; letter-spacing: -2px; background: linear-gradient(90deg, #38bdf8, #818cf8); -webkit-background-clip: text; -webkit-text-fill-color: transparent; }
    .tagline { color: #94a3b8; margin-top: 8px; font-size: 16px; }
    .container { max-width: 1000px; margin: 0 auto; padding: 32px 24px; }
    .cards { display: grid; grid-template-columns: repeat(auto-fit, minmax(140px, 1fr)); gap: 16px; margin-bottom: 32px; }
    .card { background: #1e293b; border-radius: 12px; padding: 20px; border: 1px solid #334155; }
    .card-value { font-size: 32px; font-weight: 800; }
    .card-label { font-size: 13px; color: #94a3b8; margin-top: 4px; }
    .section { background: #1e293b; border-radius: 12px; padding: 24px; margin-bottom: 24px; border: 1px solid #334155; }
    .section-title { font-size: 18px; font-weight: 700; margin-bottom: 16px; color: #f1f5f9; }
    table { width: 100%; border-collapse: collapse; }
    th { background: #0f172a; padding: 10px 12px; text-align: left; font-size: 13px; color: #94a3b8; }
    tr:nth-child(even) { background: #0f172a33; }
    .footer { text-align: center; color: #475569; font-size: 13px; padding: 32px; }
    .bleed { background: linear-gradient(135deg, #7f1d1d, #991b1b); border-radius: 12px; padding: 20px 24px; margin-bottom: 24px; border: 1px solid #ef4444; }
    .bleed-amount { font-size: 36px; font-weight: 900; color: #fca5a5; }
  </style>
</head>
<body>
  <div class="header">
    <div class="logo">CLAWCULATOR</div>
    <div class="tagline">Your friendly penny pincher. · 100% offline · Zero AI · Pure deterministic logic</div>
    <div style="color:#64748b; font-size:13px; margin-top:12px">Report generated: ${new Date(analysis.scannedAt).toLocaleString()}</div>
  </div>

  <div class="container">
    ${bleed > 0 ? `
    <div class="bleed">
      <div style="color:#fca5a5; font-size:14px; font-weight:600; margin-bottom:4px">⚠️ ESTIMATED MONTHLY COST EXPOSURE</div>
      <div class="bleed-amount">$${bleed.toFixed(2)}/month</div>
      <div style="color:#fca5a5; font-size:14px; margin-top:4px">Based on current config — fix the critical issues below to stop the bleed</div>
    </div>` : `
    <div style="background:#14532d; border-radius:12px; padding:20px 24px; margin-bottom:24px; border:1px solid #22c55e">
      <div style="color:#86efac; font-size:18px; font-weight:700">✅ No significant cost bleed detected</div>
    </div>`}

    ${summary.totalRealCost > 0 ? `
    <div class="section" style="border:1px solid #f59e0b;">
      <div class="section-title" style="color:#f59e0b;">💰 Actual API Spend (from transcripts)</div>
      <div style="display:flex; gap:32px; flex-wrap:wrap; align-items:center;">
        ${summary.todayCost > 0 ? `
        <div>
          <div style="font-size:36px; font-weight:900; color:#ef4444;">$${summary.todayCost.toFixed(2)}</div>
          <div style="font-size:13px; color:#94a3b8; margin-top:4px;">Today's spend</div>
        </div>` : ''}
        <div>
          <div style="font-size:36px; font-weight:900; color:#fbbf24;">$${summary.totalRealCost.toFixed(2)}</div>
          <div style="font-size:13px; color:#94a3b8; margin-top:4px;">All-time total (API-reported)</div>
        </div>
        ${summary.totalEstimatedCost > 0 && summary.totalRealCost > summary.totalEstimatedCost * 1.1 ? `
        <div style="background:#451a03; padding:10px 16px; border-radius:8px; border:1px solid #92400e;">
          <div style="font-size:14px; color:#fbbf24; font-weight:600;">${(summary.totalRealCost / summary.totalEstimatedCost).toFixed(1)}x higher than sessions.json estimate</div>
          <div style="font-size:12px; color:#d97706; margin-top:2px;">sessions.json: $${summary.totalEstimatedCost.toFixed(4)} — missing cache token costs</div>
        </div>` : ''}
      </div>
      ${summary.totalCacheRead > 0 || summary.totalCacheWrite > 0 ? `
      <div style="margin-top:16px; display:flex; gap:24px; flex-wrap:wrap;">
        ${summary.totalCacheWrite > 0 ? `<div style="color:#f97316; font-size:14px;">📝 Cache writes: <strong>${summary.totalCacheWrite.toLocaleString()}</strong> tokens</div>` : ''}
        ${summary.totalCacheRead > 0 ? `<div style="color:#22c55e; font-size:14px;">📖 Cache reads: <strong>${summary.totalCacheRead.toLocaleString()}</strong> tokens</div>` : ''}
      </div>` : ''}
    </div>` : ''}

    <div class="cards">
      <div class="card">
        <div class="card-value" style="color:#ef4444">${summary.critical}</div>
        <div class="card-label">🔴 Critical</div>
      </div>
      <div class="card">
        <div class="card-value" style="color:#f97316">${summary.high}</div>
        <div class="card-label">🟠 High</div>
      </div>
      <div class="card">
        <div class="card-value" style="color:#eab308">${summary.medium}</div>
        <div class="card-label">🟡 Medium</div>
      </div>
      <div class="card">
        <div class="card-value" style="color:#06b6d4">${summary.low || 0}</div>
        <div class="card-label">🔵 Low</div>
      </div>
      <div class="card">
        <div class="card-value" style="color:#22c55e">${summary.info}</div>
        <div class="card-label">✅ OK</div>
      </div>
      <div class="card">
        <div class="card-value" style="color:#38bdf8">${summary.sessionsAnalyzed}</div>
        <div class="card-label">Sessions Analyzed</div>
      </div>
      <div class="card">
        <div class="card-value" style="color:#818cf8">${(summary.totalTokensFound || 0).toLocaleString()}</div>
        <div class="card-label">Total Tokens Found</div>
      </div>
    </div>

    <div class="section">
      <div class="section-title">Findings</div>
      <div style="color:#0f172a">
        ${findingCards}
      </div>
    </div>

    ${activeRows ? `
    <div class="section">
      <div class="section-title">Active Sessions</div>
      <div style="overflow-x:auto">
        <table>
          <thead><tr><th>Session</th><th>Model</th><th style="text-align:right">Tokens</th><th style="text-align:right">Total Cost</th><th style="text-align:right">$/day</th><th>Last Active</th></tr></thead>
          <tbody>${activeRows}</tbody>
        </table>
      </div>
      <div style="margin-top:8px; font-size:12px; color:#64748b;">● = cost from transcript · ○ = estimated · ⚠️ = orphaned</div>
    </div>` : ''}

    ${untrackedRows ? `
    <div class="section" style="border:1px solid #475569;">
      <div class="section-title" style="color:#94a3b8;">📦 Historical Sessions — $${untrackedTotal.toFixed(2)} total</div>
      <div style="font-size:13px; color:#64748b; margin-bottom:12px;">Old/deleted sessions still on disk. Top 10 by cost.</div>
      <div style="overflow-x:auto">
        <table>
          <thead><tr><th>Session</th><th>Model</th><th style="text-align:right">Tokens</th><th style="text-align:right">Total Cost</th><th style="text-align:right">$/day</th><th>Last Active</th></tr></thead>
          <tbody>${untrackedRows}</tbody>
        </table>
      </div>
      ${untrackedHidden > 0 ? `<div style="margin-top:8px; font-size:12px; color:#64748b;">+ ${untrackedHidden} more not shown</div>` : ''}
    </div>` : ''}

    ${burnSummary}

    <div class="section">
      <div class="section-title">Summary</div>
      <div style="display:grid; grid-template-columns:repeat(auto-fit, minmax(200px, 1fr)); gap:16px;">
        <div>
          <div style="font-size:13px; color:#94a3b8;">Severity Breakdown</div>
          <div style="margin-top:4px;">🔴 ${summary.critical} critical · 🟠 ${summary.high} high · 🟡 ${summary.medium} medium · 🔵 ${summary.low||0} low · ✅ ${summary.info} ok</div>
        </div>
        <div>
          <div style="font-size:13px; color:#94a3b8;">Sessions Analyzed</div>
          <div style="margin-top:4px; font-size:18px; font-weight:700;">${summary.sessionsAnalyzed}</div>
        </div>
        <div>
          <div style="font-size:13px; color:#94a3b8;">Total Tokens</div>
          <div style="margin-top:4px; font-size:18px; font-weight:700;">${(summary.totalTokensFound||0).toLocaleString()}</div>
        </div>
        ${summary.todayCost > 0 ? `<div>
          <div style="font-size:13px; color:#94a3b8;">Today's Spend</div>
          <div style="margin-top:4px; font-size:18px; font-weight:700; color:#ef4444;">$${summary.todayCost.toFixed(2)}</div>
        </div>` : ''}
        <div>
          <div style="font-size:13px; color:#94a3b8;">All-Time Spend</div>
          <div style="margin-top:4px; font-size:18px; font-weight:700; color:#fbbf24;">$${summary.totalRealCost.toFixed(2)}</div>
        </div>
        ${summary.totalCacheRead > 0 ? `<div>
          <div style="font-size:13px; color:#94a3b8;">Cache Tokens</div>
          <div style="margin-top:4px; font-size:14px;">${(summary.totalCacheRead||0).toLocaleString()} read<br/>${(summary.totalCacheWrite||0).toLocaleString()} write</div>
        </div>` : ''}
        ${summary.totalEstimatedCost > 0 && summary.totalRealCost > summary.totalEstimatedCost * 1.1 ? `<div>
          <div style="font-size:13px; color:#94a3b8;">sessions.json Gap</div>
          <div style="margin-top:4px; font-size:14px; color:#f59e0b;">Estimate: $${summary.totalEstimatedCost.toFixed(4)}<br/>${(summary.totalRealCost / summary.totalEstimatedCost).toFixed(1)}x under-reported</div>
        </div>` : ''}
        ${bleed > 0 ? `<div>
          <div style="font-size:13px; color:#94a3b8;">Monthly Bleed</div>
          <div style="margin-top:4px; font-size:18px; font-weight:700; color:#ef4444;">$${bleed.toFixed(2)}/mo</div>
        </div>` : ''}
      </div>
    </div>

    ${(() => {
      const wins = findings.filter(f => f.fix && f.severity !== 'info');
      if (!wins.length) return '';
      return `
    <div class="section" style="border:1px solid #22c55e;">
      <div class="section-title" style="color:#22c55e;">⚡ Quick Wins</div>
      ${wins.slice(0, 5).map((f, i) => `
        <div style="margin-bottom:16px;">
          <div style="color:#22c55e; font-weight:600; margin-bottom:4px;">${i+1}. ${f.fix}</div>
          ${f.command ? `<div style="background:#0f172a; color:#38bdf8; font-family:monospace; font-size:13px; padding:10px 14px; border-radius:6px; white-space:pre-wrap; border:1px solid #1e293b;"><span style="color:#6b7280; user-select:none;">$ </span>${f.command}</div>` : ''}
        </div>
      `).join('')}
    </div>`;
    })()}

  </div>
  <div class="footer">
    Clawculator · github.com/echoudhry/clawculator · Your friendly penny pincher.
  </div>
</body>
</html>`;

  if (!outPath) {
    outPath = path.join(process.cwd(), `clawculator-report-${Date.now()}.html`);
  }
  fs.writeFileSync(outPath, html, 'utf8');
  return outPath;
}

module.exports = { generateHTMLReport };

File: .skills/openclaw-skills/skills/zyuanjia/webnovel-writer/scripts/html_report.py
#!/usr/bin/env python3
"""检测报告生成器 - 将检测结果渲染为可视化 HTML 报告

用法:
    html_report <正文目录> [--output <路径>] [--title <标题>]
"""

import html
import json
import os
import re
import sys
import time
from pathlib import Path
from datetime import datetime
from typing import List, Tuple

SCRIPTS_DIR = Path(__file__).parent


def _read(path):
    with open(path, "r", encoding="utf-8") as f:
        return f.read()


def run_check(script_name: str, args: list) -> str:
    """运行检测脚本，返回stdout"""
    script = SCRIPTS_DIR / f"{script_name}.py"
    if not script.exists():
        return ""
    import subprocess
    result = subprocess.run(
        [sys.executable, str(script)] + args,
        capture_output=True, text=True, timeout=60
    )
    return result.stdout


def list_chapters(directory: str) -> List[Tuple[int, str, str]]:
    chapters = []
    for f in sorted(Path(directory).glob("*.md")):
        m = re.search(r"第(\d+)", f.name)
        if m:
            chapters.append((int(m.group(1)), str(f), f.name))
    return chapters


def generate_html(
    novel_dir: str,
    title: str = "检测报告",
    output: str = None,
) -> str:
    """生成完整的 HTML 检测报告"""
    chapters = list_chapters(novel_dir)
    if not chapters:
        print("❌ 未找到章节文件")
        return ""

    total_chars = 0
    for _, path, _ in chapters:
        total_chars += len(_read(path))

    # 收集各检测脚本结果
    sections = []

    # 1. 一致性检查
    print("  ⏳ 一致性检查...")
    result = run_check("consistency_check", [novel_dir])
    sections.append(("一致性检查", "shield", result))

    # 2. 重复检测
    print("  ⏳ 重复检测...")
    result = run_check("repetition_check", [novel_dir])
    sections.append(("重复检测", "copy", result))

    # 3. 对话标签
    print("  ⏳ 对话标签...")
    result = run_check("dialogue_tag_check", [novel_dir])
    sections.append(("对话标签质量", "message-square", result))

    # 4. 章末钩子
    print("  ⏳ 章末钩子...")
    result = run_check("chapter_hook_check", [novel_dir])
    sections.append(("章末钩子", "anchor", result))

    # 5. 段落检测
    print("  ⏳ 段落检测...")
    result = run_check("paragraph_check", [novel_dir])
    sections.append(("段落长度", "align-left", result))

    # 6. 节奏检测
    print("  ⏳ 节奏检测...")
    result = run_check("rhythm_check", [novel_dir])
    sections.append(("节奏分析", "activity", result))

    # 构建 HTML
    now = datetime.now().strftime("%Y-%m-%d %H:%M:%S")
    section_cards = ""
    for i, (name, icon, content) in enumerate(sections):
        safe_content = html.escape(content) if content else "✅ 无问题"
        # 统计问题数
        issue_count = content.count("❌") + content.count("⚠️") + content.count("⚠")
        status_color = "#ef4444" if issue_count > 10 else "#f59e0b" if issue_count > 0 else "#10b981"
        status_text = f"{issue_count} 个问题" if issue_count > 0 else "通过"
        section_cards += f"""
        <div class="card">
            <div class="card-header">
                <span class="icon">📊</span>
                <h2>{html.escape(name)}</h2>
                <span class="badge" style="background:{status_color}">{status_text}</span>
            </div>
            <details>
                <summary>展开详情</summary>
                <pre class="output">{safe_content}</pre>
            </details>
        </div>"""

    full_html = f"""<!DOCTYPE html>
<html lang="zh-CN">
<head>
<meta charset="UTF-8">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>{html.escape(title)}</title>
<style>
    * {{ margin: 0; padding: 0; box-sizing: border-box; }}
    body {{ font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', sans-serif;
           background: #0f172a; color: #e2e8f0; padding: 2rem; }}
    .container {{ max-width: 900px; margin: 0 auto; }}
    h1 {{ font-size: 1.8rem; margin-bottom: 0.5rem; }}
    .meta {{ color: #94a3b8; margin-bottom: 2rem; }}
    .stats {{ display: grid; grid-template-columns: repeat(auto-fit, minmax(140px, 1fr));
             gap: 1rem; margin-bottom: 2rem; }}
    .stat {{ background: #1e293b; border-radius: 12px; padding: 1rem; text-align: center; }}
    .stat-value {{ font-size: 1.5rem; font-weight: bold; color: #38bdf8; }}
    .stat-label {{ font-size: 0.85rem; color: #94a3b8; margin-top: 0.25rem; }}
    .card {{ background: #1e293b; border-radius: 12px; margin-bottom: 1rem;
            overflow: hidden; border: 1px solid #334155; }}
    .card-header {{ display: flex; align-items: center; gap: 0.75rem;
                   padding: 1rem 1.25rem; cursor: pointer; }}
    .card-header h2 {{ font-size: 1.1rem; flex: 1; }}
    .icon {{ font-size: 1.2rem; }}
    .badge {{ padding: 0.2rem 0.6rem; border-radius: 999px; font-size: 0.75rem;
              color: white; font-weight: bold; }}
    summary {{ cursor: pointer; padding: 0.75rem 1.25rem; color: #38bdf8;
              border-top: 1px solid #334155; font-size: 0.9rem; }}
    .output {{ padding: 1rem 1.25rem; font-size: 0.8rem; line-height: 1.6;
               overflow-x: auto; white-space: pre-wrap; color: #cbd5e1; max-height: 500px; }}
    .footer {{ text-align: center; color: #475569; margin-top: 2rem; font-size: 0.8rem; }}
</style>
</head>
<body>
<div class="container">
    <h1>📖 {html.escape(title)}</h1>
    <div class="meta">生成时间: {now} | Novel Writer v3.2</div>

    <div class="stats">
        <div class="stat">
            <div class="stat-value">{len(chapters)}</div>
            <div class="stat-label">总章数</div>
        </div>
        <div class="stat">
            <div class="stat-value">{total_chars:,}</div>
            <div class="stat-label">总字数</div>
        </div>
        <div class="stat">
            <div class="stat-value">{total_chars // len(chapters) if chapters else 0:,}</div>
            <div class="stat-label">平均字数/章</div>
        </div>
        <div class="stat">
            <div class="stat-value">{6}</div>
            <div class="stat-label">检测维度</div>
        </div>
    </div>

    {section_cards}

    <div class="footer">
        Generated by <a href="https://github.com/ZYuanJia/novel-writer" style="color:#38bdf8">Novel Writer</a>
    </div>
</div>
</body>
</html>"""

    if output:
        Path(output).parent.mkdir(parents=True, exist_ok=True)
        with open(output, "w", encoding="utf-8") as f:
            f.write(full_html)
        print(f"\n✅ 报告已保存: {output}")
    else:
        default_path = str(Path(novel_dir).parent / f"检测报告_{datetime.now().strftime('%Y%m%d_%H%M%S')}.html")
        with open(default_path, "w", encoding="utf-8") as f:
            f.write(full_html)
        print(f"\n✅ 报告已保存: {default_path}")

    return full_html


def main():
    if len(sys.argv) < 2:
        print("📖 HTML 检测报告生成器")
        print("用法: html_report <正文目录> [--output <路径>] [--title <标题>]")
        sys.exit(0)

    novel_dir = sys.argv[1]
    output = None
    title = "小说质量检测报告"

    for i, arg in enumerate(sys.argv):
        if arg == "--output" and i + 1 < len(sys.argv):
            output = sys.argv[i + 1]
        if arg == "--title" and i + 1 < len(sys.argv):
            title = sys.argv[i + 1]

    if not Path(novel_dir).is_dir():
        print(f"❌ 目录不存在: {novel_dir}")
        sys.exit(1)

    print(f"🔍 生成检测报告...\n")
    generate_html(novel_dir, title, output)


if __name__ == "__main__":
    main()
